File: backend/src/main/java/com/ort/dbflute/exentity/Skills.java
package com.ort.dbflute.exentity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import com.ort.dbflute.allcommon.CDef;

public class Skills {

    public List<CDef.Skill> list;

    public Skills(List<CDef.Skill> list) {
        this.list = list;
    }

    public static Skills of() {
        return new Skills(CDef.Skill.listAll());
    }

    public Skills filterBy(Predicate<CDef.Skill> predicate) {
        return new Skills(this.list.stream().filter(predicate).collect(Collectors.toList()));
    }

    public Skills filterByCamp(CDef.Camp camp) {
        return this.filterBy(s -> s.campCode().equals(camp.code()));
    }

    public Skills filterNotSomeone() {
        return this.filterBy(s -> !s.isSomeoneSkill());
    }

    public Skills sortedBy(Comparator<CDef.Skill> comparator) {
        return new Skills(this.list.stream().sorted(comparator).collect(Collectors.toList()));
    }

    public Skills sortedByDispOrder() {
        return this.sortedBy((s1, s2) -> Integer.parseInt(s1.order()) - Integer.parseInt(s2.order()));
    }

    public CDef.Skill getRandom() {
        if (this.list.isEmpty())
            throw new IllegalStateException("skill_list is empty.");
        List<CDef.Skill> newList = new ArrayList<>(this.list);
        Collections.shuffle(newList);
        return newList.get(0);
    }
}

File: src/main/kotlin/com/ort/app/domain/model/skill/Skills.kt
package com.ort.app.domain.model.skill

import com.ort.dbflute.allcommon.CDef

data class Skills(val list: List<Skill>) {
    companion object {
        fun all(): Skills = Skills(
            list = CDef.Skill.listAll().sortedBy { it.order().toInt() }.map { Skill(it) }
        )

        fun someones(): Skills = Skills(
            list = CDef.Skill.listOfSomeoneSkill().sortedBy { it.order().toInt() }.map { Skill(it) }
        )

        private val notRevivableSkills = listOf(CDef.Skill.同棲者, CDef.Skill.恋人)
        val openSkills = listOf(CDef.Skill.勇者, CDef.Skill.絶対人狼)

        // 転生で割り当て可能な役職
        fun revivables(): Skills = Skills(
            list = all().filterNotSomeone().list.filterNot { notRevivableSkills.contains(it.toCdef()) }
        )

        // 役職希望で選択できる役職
        fun requestables(): Skills = Skills(
            list = all().list.filter { it.isRequestable() }
        )

        // 編成に含めることができる役職
        fun organizables(): Skills = Skills(
            list = all().filterNotSomeone().list.filter { it.isRequestable() }
        )

        fun getSkillListStr(): String =
            organizables().list.joinToString(separator = " / ") { "${it.name}:${it.shortName}" }
    }

    fun filterByCamp(camp: CDef.Camp): Skills = copy(list = list.filter { it.toCdef().campCode() == camp.code() })

    fun filterNotSomeone(): Skills = copy(list = list.filterNot { it.toCdef().isSomeoneSkill })

    fun filter(predicate: (Skill) -> Boolean): Skills = copy(list = list.filter { predicate(it) })
}